File: python/confident_agent/handler/loader.py
import importlib.util
import sys
from pathlib import Path
from typing import Callable

from .decorator import _registered


class HandlerError(Exception):
    pass


def load_handler(path: str) -> Callable:
    file = Path(path).expanduser().resolve()
    if not file.is_file():
        raise HandlerError(f"handler file not found: {file}")

    parent = str(file.parent)
    if parent not in sys.path:
        sys.path.insert(0, parent)

    _registered.clear()

    spec = importlib.util.spec_from_file_location("confident_handler_module", file)
    if spec is None or spec.loader is None:
        raise HandlerError(f"could not import handler file: {file}")
    module = importlib.util.module_from_spec(spec)
    try:
        spec.loader.exec_module(module)
    except Exception as e:
        raise HandlerError(f"handler file failed to import: {e}") from e

    if len(_registered) == 0:
        raise HandlerError(
            f"no @handler function found in {file.name} — decorate exactly one "
            "function with @handler from confident_agent"
        )
    if len(_registered) > 1:
        names = ", ".join(fn.__name__ for fn in _registered)
        raise HandlerError(
            f"multiple @handler functions found in {file.name} ({names}) — "
            "exactly one is allowed per agent"
        )
    return _registered[0]

File: typescript/src/handler/loader.ts
import { statSync } from "node:fs";
import { homedir } from "node:os";
import { basename, join, resolve } from "node:path";
import { pathToFileURL } from "node:url";

import { registered, type HandlerFn } from "./decorator.js";

export class HandlerError extends Error {}

export async function loadHandler(path: string): Promise<HandlerFn> {
  const expanded = path.startsWith("~") ? join(homedir(), path.slice(1)) : path;
  const file = resolve(expanded);
  if (!isFile(file)) {
    throw new HandlerError(`handler file not found: ${file}`);
  }

  registered.length = 0;

  try {
    await import(`${pathToFileURL(file).href}?t=${Date.now()}`);
  } catch (e) {
    throw new HandlerError(
      `handler file failed to import: ${e instanceof Error ? e.message : String(e)}`
    );
  }

  if (registered.length === 0) {
    throw new HandlerError(
      `no handler function found in ${basename(file)} — register exactly one ` +
        "function with handler() from confident-agent"
    );
  }
  if (registered.length > 1) {
    const names = registered.map((fn) => fn.name).join(", ");
    throw new HandlerError(
      `multiple handler functions found in ${basename(file)} (${names}) — ` +
        "exactly one is allowed per agent"
    );
  }
  return registered[0];
}

function isFile(file: string): boolean {
  try {
    return statSync(file).isFile();
  } catch {
    return false;
  }
}
